### ISSUE TITLE
Taiko no Tatsujin: Get 2-4 BACKUP MEMORY ERROR 2 error

### ISSUE THREAD

**Issue by gkr-758** [RPCS3/rpcs3#13390]
### Quick summary

When I start the first version of Taiko no Tatsujin, I get the i/o error "2-4 BACKUP MEMORY ERROR 2".

### Details

![B2SKBIo](https://user-images.githubusercontent.com/85799458/218311859-21272662-e79e-4ec1-98c1-b8b89a3fa51b.png)

### Attach a log file

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/10716072/RPCS3.log.gz)

### System configuration

Emu: RPCS3 v0.0.26-14704-2f7a2d51 Alpha | master | FW 4.70 (GEX)
CPU: AMD Ryzen 5 3500 | 16.0 GB RAM
GPU: NVIDIA GeForce GTX 1650

**Comment by Megamouse** [comment]
reset your configuration to default and upload a new log...

**Comment by Megamouse** [comment]
also disable all patches

**Comment by gkr-758** [comment]
> also disable all patches

If the patch is not applied, the "19-2 SYSTEM ERROR 2 (SECURITY CHECK 2)" check will not pass and a different error will occur

**Comment by Megamouse** [comment]
well then tell us the proper error.... wtf

**Comment by gkr-758** [comment]
so I will try again tonight

**Comment by gkr-758** [comment]
> 
This is a common error in other versions and can be resolved by patching, so it is not that important

**Comment by gkr-758** [comment]
"19-2 SYSTEM ERROR 2 (SECURITY CHECK 2)" exists in other versions and is a simple error that can be worked around with a patch.
However, "2-4 BACKUP MEMORY ERROR 2" is an I/O problem and cannot be patched around.
So I need to fix I/O
("2-4 BACKUP MEMORY ERROR 2" occurs only in the first version of Taiko no Tatsujin, not in later versions)

**Comment by brian218** [comment]
I am pretty sure the patch is no longer needed for 19-2 SYSTEM ERROR 2.
Also, do let us know the behavior of real hardware testing, and clarify how you would do when the BACKUP MEMORY ERROR occurs on real hardware.

**Comment by gkr-758** [comment]
I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.

**Comment by brian218** [comment]
> I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.

I suggest that you get in touch with this guy to ask him about how he solved the 2-4 BACKUP MEMORY ERROR on his real hardware. https://hajime0512.hatenablog.com/

**Comment by gkr-758** [comment]
> > I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.
> 
> I suggest that you get in touch with this guy to ask him about how he solved the 2-4 BACKUP MEMORY ERROR on his real hardware. https://hajime0512.hatenablog.com/

Real hardware and emulator specifications are different, so there is no point in asking questions.

**Comment by brian218** [comment]
> > > I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.
> > 
> > 
> > I suggest that you get in touch with this guy to ask him about how he solved the 2-4 BACKUP MEMORY ERROR on his real hardware. https://hajime0512.hatenablog.com/
> 
> Real hardware and emulator specifications are different, so there is no point in asking questions.

Since emulation is basically reflecting the real hardware's behavior (at least in this case - USIO emulation), you will need to clarify the real hardware's behavior before we can figure out how to solve this problem for the emulator.

**Comment by gkr-758** [comment]
> > > > I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.
> > > 
> > > 
> > > I suggest that you get in touch with this guy to ask him about how he solved the 2-4 BACKUP MEMORY ERROR on his real hardware. https://hajime0512.hatenablog.com/
> > 
> > 
> > Real hardware and emulator specifications are different, so there is no point in asking questions.
> 
> Since emulation is basically reflecting the real hardware's behavior (at least in this case - USIO emulation), you will need to clarify the real hardware's behavior before we can figure out how to solve this problem for the emulator.

He wrote in his blog that the error was caused by using I/O that is used by another game, so he probably got the I/O for Taiko no Tatsujin and used that.

**Comment by brian218** [comment]
> > > > > I don't have real hardware so I don't know, but the instruction manual says to replace the I/O board.
> > > > 
> > > > 
> > > > I suggest that you get in touch with this guy to ask him about how he solved the 2-4 BACKUP MEMORY ERROR on his real hardware. https://hajime0512.hatenablog.com/
> > > 
> > > 
> > > Real hardware and emulator specifications are different, so there is no point in asking questions.
> > 
> > 
> > Since emulation is basically reflecting the real hardware's behavior (at least in this case - USIO emulation), you will need to clarify the real hardware's behavior before we can figure out how to solve this problem for the emulator.
> 
> He wrote in his blog that the error was caused by using I/O that is used by another game, so he probably got the I/O for Taiko no Tatsujin and used that.

In addition, there doesn't seem to be any problem in the code of the current implementation.
The current USIO emulation implementation is intended for Taiko Green, but there are chances that Taiko no Tatsujin 2011 may expect different data or data format from the USIO as it is the oldest version in the game series, and the data or the data format might have been changed by the updates of the USIO firmware in later Taiko versions.
We can only figure out the solution by real hardware sniffing, since we are not the manufacturer of the USIO, nor do we have the blueprint or source code of that.

**Comment by ChwaraeTeg** [comment]
Does this need a retest after #14248 was merged?

**Comment by brian218** [comment]
> Does this need a retest after #14248 was merged?

Unfortunately no.
There's another problem (F {PPU[0x1000000] Thread (main_thread) [0x0002850c]} VM: Access violation reading location 0x3 (unmapped memory)).

**Comment by brian218** [comment]
Tested on my real hardware, the game crashed on real hardware when it detected my USIO of blank backup memory.
It's the same behaviour on both real hardware and the emulator (with blank usiobackup.bin) so it's not the emulator's fault.

**Comment by brian218** [comment]
I think this can be closed unless further discovery regarding the emulator/hardware behaviour difference is reported.

**Comment by brian218** [comment]
> I don't understand.. shouldn't this work with the emulated usio?

The game even cannot work with real USIO, why should we expect it to work with the emulated USIO which emulates the real USIO?
The game either shows 2-4 BACKUP MEMORY ERROR with used USIO like the OP or crashes with new USIO.

**Comment by Megamouse** [comment]
So the game does not work at all on real hardware?

**Comment by brian218** [comment]
> So the game does not work at all on real hardware?

Yep.
I believe an USIO with older firmware is needed.

**Comment by Megamouse** [comment]
If it works with an older usio, it's still fixable through emulation though.

**Comment by brian218** [comment]
> If it works with an older usio, it's still fixable through emulation though.

I cannot confirm that. It's just my assumption.
If that's the case, emulating an older USIO can break the compatibility with the other games which have been confirmed to be able to run through current emulation.

**Comment by Megamouse** [comment]
Not really if it's an option. I'm not really sure how to solve this.

**Comment by brian218** [comment]
> Not really if it's an option. I'm not really sure how to solve this.

So it has no solution.
I even tried the game with various USJ/USIO boards I have.
